A79-non-per-call-metrics-architecture.md

```diff
@@ -0,0 +1,405 @@
+corresponding metrics data (timeseries).
+
+New metrics should start as experimental. Once they are stable, they can
+optionally be made on-by-default. The distinction between on-by-default and
+off-by-default metrics allows gRPC to support metrics that may be desirable in
+some cases but too expensive to collect by default.
+
+### Changes to gRPC OpenTelemetry Plugin
+
+Optional Labels - Users should be able to configure gRPC OpenTelemetry plugins
+to record the additional optional labels. One suggested approach of implementing
+this is for the plugin instance to store a bit set for each instrument where
+each bit denotes whether the plugin should record the label in the corresponding
+position of the set of optional labels.
+
+API to enable/disable metrics - Users should be able to override the default
+behavior.
+
+Channel Scoped OpenTelemetry Plugins - API should be provided based on the
+approach chosen from the `Scoping Stats Plugins for Channels` section.
+
+#### C++
+
+The following API will be added to `OpenTelemetryPluginBuilder` introduced in
+[A66: OpenTelemetry Metrics](A66-otel-stats.md).
+
+```c++
+class OpenTelemetryPluginBuilder {
+  // Records \a optional_label_key on all metrics that provide it.
+  void AddOptionalLabel(absl::string_view optional_label_key);
+  // Set scope filter to choose which channels are recorded by this plugin.
+  // Server-side recording remains unaffected.
+  OpenTelemetryPluginBuilder& SetChannelScope(
+      absl::AnyInvocable<bool(ChannelScope /*scope*/) const> scope_filter);
+  // Enable/Disable metrics.
+  OpenTelemetryPluginBuilder& EnableMetrics(
+      absl::flat_hash_set<absl::string_view> metrics);
+  OpenTelemetryPluginBuilder& DisableMetrics(
+      absl::flat_hash_set<absl::string_view> metrics);
+  OpenTelemetryPluginBuilder& DisableAllMetrics();
+};
+```
+
+#### Java
+
+(To be filled)
+
+#### Go
+
+(To be filled)
+
+## Rationale
+
+An alternate approach to record non-per-call metrics would be to extend the
+existing event based per-attempt/call architecture from
+[A66: OpenTelemetry Metrics](A66-otel-stats.md). Following the footsteps of a
+`CallTracer`, a `ChannelTracer` could be used to record events from various gRPC
+components. Instead of registering an instrument descriptor for metrics they
+want to record, components would instead have call-outs to the `ChannelTracer`
+on various interesting events. This approach has two major issues.
+
+1.  The stats plugin defines the metrics schema and every metric would need to
+    be implemented in every supported stats plugin. The stats plugin
+    implementation of the `ChannelTracer` would also need to know about events
+    from all possible gRPC components so that it can record metrics for those
+    components. This would not work for third-party plugins. (Note that one
+    major reason the `CallTracer` architecture was chosen for per-attempt/call
+    metrics was for legacy reasons. It was already implemented in various
```

Comment:
I don't think the legacy reasons are the main issue here; I think the main reason is performance.  If it weren't for the performance issues, I think we would have switched the per-call metrics in A66 to also be implemented using the non-per-call framework described in this gRFC, since that would have kept gRPC (as opposed to the stats plugin impl) in control of the schema.

Reply:
For our opentelemetry implementation, we don't have batching support for opentelemetry anyway, and thinking about it, if we wanted to we could batch multiple measurements with the non-per-call metrics architecture as well.

The legacy reasons which resulted in different metrics between the different implementations seems to be the major issue to me.

Reply:
Amended the text.